nelsk-kore/lib/kore/comm/platform_adapter.rb

```diff
@@ -0,0 +1,74 @@
+module Kore
+  module Comm
+    class PlatformAdapter < Kore::Machinery::Base
+      attr_accessor :engine
+      TRIGGER_PREFIX = '!'
+
+      def message_received(identity, raw)
+        self.log.debug "PlatformAdapter#message_received"
+        self.log.debug "  identity: #{identity}"
+        self.log.debug "  raw: #{raw}"
+
+        if self.is_cmd(raw)
+          self.log.debug "Command matched, parsing..."
+          begin
+            m = self.parse(raw)
+            m.originator = {
+              identity: identity,
+              platform: self.name
+            }
+          rescue Exception => e
+            self.log.warn "Error occurred while parsing message"
+            self.log.warn raw
+            self.log.warn e.message
+          else
+            @engine.route_ingress(m)
+          end
+        end
+      end
+
+      #############################################################
+      # Interface methods to be overriden
+      #############################################################
+      def send_message(emsg)
+        raise "platform adapter #{self.name} must implement \"send_message\" method"
+      end
+
+      def name
+        raise 'platform adapters must implement "name" method'
+      end
+
+      # Must be non-blocking!
+      def listen
+        raise "platform adapter #{self.name} must implement \"listen\" method"
+      end
+      #############################################################
+
+      def is_cmd(raw)
+        !!(Regexp.new("^#{TRIGGER_PREFIX}\\S*($| )").match(raw))
+      end
+
+      def parse(raw)
```

Comment:
To me, the input type of this plugin is very unobvious (I *think* its an array?).  We should really get some documentation going if we are going to have more than one person hacking on this at a time and want community contributors. This doesn't need to be done now but as soon as work starts on the actual implementation, we should have a documentation guideline in place.

Reply:
Very good point. parse is defined by the concrete adapter, because the messages are going to very based on where they're coming from. I'd like to follow up with an adjustment that demo's that truth.